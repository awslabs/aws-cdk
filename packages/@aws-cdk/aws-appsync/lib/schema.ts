import { IBucket } from '@aws-cdk/aws-s3';
import { CfnGraphQLSchema } from './appsync.generated';
import { GraphqlApi } from './graphqlapi';
import { AssetSchema, CodeSchema, S3Schema } from './private/schema-implementation';
import { IIntermediateType } from './schema-base';
import { ResolvableField } from './schema-field';
import { ObjectType } from './schema-intermediate';

/**
 * The Schema for a GraphQL Api
 *
 * If no options are configured, schema will be generated
 * code-first.
 */
export abstract class Schema {
  /**
   * Generate a Schema from file
   *
   * @returns `Schema` with immutable schema defintion
   * @param filePath the file path of the schema file
   */
  public static fromAsset(filePath: string): Schema {
    return new AssetSchema(filePath);
  }

  /**
   * Point GraphQL Schema to a S3 Location
   *
   * @returns `Schema` with immutable schema defintion
   * @param bucket the s3 bucket that holds the GraphQL Schema
   * @param key the object key for the schema
   */
  public static fromBucket(bucket: IBucket, key: string): Schema {
    return new S3Schema(bucket, key);
  }

  /**
   * Generate a Schema that can be edited through a code-first
   * approach
   *
   * @returns `Schema` with mutable schema defintion
   */
  public static fromCode(): Schema {
    return new CodeSchema();
  }

  /**
   * The definition for this schema
   */
  public readonly abstract definition: string;

  /**
   * Called when the GraphQL Api is initialized to allow this object to bind
   * to the stack.
   */
  public abstract bind(api: GraphqlApi): CfnGraphQLSchema;

  /**
   * Escape hatch to add to Schema as desired. Will always result
   * in a newline.
   */
  public abstract addToSchema(addition: string, delimiter?: string): void;
  /**
   * Add a query field to the schema's Query. CDK will create an
   * Object Type called 'Query'. For example,
   *
   * type Query {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addQuery(fieldName: string, field: ResolvableField): ObjectType;

  /**
   * Add a mutation field to the schema's Mutation. CDK will create an
   * Object Type called 'Mutation'. For example,
   *
   * type Mutation {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addMutation(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add a subscription field to the schema's Subscription. CDK will create an
   * Object Type called 'Subscription'. For example,
   *
   * type Subscription {
   *   fieldName: Field.returnType
   * }
   */
  public abstract addSubscription(_fieldName: string, _field: ResolvableField): ObjectType;

  /**
   * Add type to the schema
   */
  public abstract addType(_type: IIntermediateType): IIntermediateType;
}