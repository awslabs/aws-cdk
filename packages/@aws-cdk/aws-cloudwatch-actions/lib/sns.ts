import * as cloudwatch from '@aws-cdk/aws-cloudwatch';
import { ServicePrincipal } from '@aws-cdk/aws-iam';
import * as sns from '@aws-cdk/aws-sns';
import { Construct } from '@aws-cdk/core';

/**
 * Use an SNS topic as an alarm action
 */
export class SnsAction implements cloudwatch.IAlarmAction {
  constructor(private readonly topic: sns.ITopic) {
  }

  /**
   * Returns an alarm action configuration to use an SNS topic as an alarm action
   */
  public bind(_scope: Construct, _alarm: cloudwatch.IAlarm): cloudwatch.AlarmActionConfig {
    if (this.topic.masterKey) {
      this.topic.masterKey.grantEncryptDecrypt(new ServicePrincipal('cloudwatch.amazonaws.com'));
    }
    return { alarmActionArn: this.topic.topicArn };
  }
}
