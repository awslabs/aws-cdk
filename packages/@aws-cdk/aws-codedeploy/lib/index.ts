export * from './rollback-config';
export * from './ecs';
export * from './lambda';
export * from './server';
export * from './trigger-configuration';

// AWS::CodeDeploy CloudFormation Resources:
export * from './codedeploy.generated';
