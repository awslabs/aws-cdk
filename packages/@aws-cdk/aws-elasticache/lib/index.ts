// AWS::ElastiCache CloudFormation Resources:
export * from './elasticache.generated';
export * from './parameter-group';
