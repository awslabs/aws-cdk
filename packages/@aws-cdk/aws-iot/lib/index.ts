export * from './topic-rule';
export * from './topic-rule-action';
// AWS::IoT CloudFormation Resources:
export * from './iot.generated';
