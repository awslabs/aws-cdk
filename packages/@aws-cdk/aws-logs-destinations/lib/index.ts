export * from './lambda';
export * from './kinesis';
export * from './kinesisfirehose';
