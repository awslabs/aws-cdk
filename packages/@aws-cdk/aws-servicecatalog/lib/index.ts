export * from './common';
export * from './constraints';
export * from './cloudformation-template';
export * from './portfolio';
export * from './product';
export * from './tagoptions';

// AWS::ServiceCatalog CloudFormation Resources:
export * from './servicecatalog.generated';
