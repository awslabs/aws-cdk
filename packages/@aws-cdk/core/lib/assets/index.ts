export * from './asset-staging';
export * from './bundling';
export * from './common';
export * from './file-asset';
export * from './image-asset';
