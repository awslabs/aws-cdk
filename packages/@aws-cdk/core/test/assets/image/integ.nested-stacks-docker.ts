import * as path from 'path';
import * as cxapi from '@aws-cdk/cx-api';
import { Construct } from 'constructs';
import { App, CfnOutput, CfnResource, ImageAsset, NestedStack, NestedStackProps, Stack, StackProps } from '../../../lib';

class TheNestedStack extends NestedStack {
  constructor(scope: Construct, id: string, props?: NestedStackProps) {
    super(scope, id, props);

    const asset = new ImageAsset(this, 'my-image', {
      directory: path.join(__dirname, 'demo-image'),
    });

    new CfnResource(this, 'User', { type: 'AWS::IAM::User' });
    new CfnOutput(this, 'output', { value: asset.imageUri });
  }
}

class TheParentStack extends Stack {
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props);

    new TheNestedStack(this, 'nested-stack-with-image');
  }
}

const app = new App({
  context: {
    [cxapi.DOCKER_IGNORE_SUPPORT]: true,
  },
});
new TheParentStack(app, 'nested-stacks-docker');
app.synth();
