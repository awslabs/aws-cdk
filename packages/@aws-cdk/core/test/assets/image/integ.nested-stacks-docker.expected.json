{
  "Resources": {
    "nestedstackwithimageNestedStacknestedstackwithimageNestedStackResourceDF784FD5": {
      "Type": "AWS::CloudFormation::Stack",
      "Properties": {
        "TemplateURL": {
          "Fn::Join": [
            "",
            [
              "https://s3.",
              {
                "Ref": "AWS::Region"
              },
              ".",
              {
                "Ref": "AWS::URLSuffix"
              },
              "/",
              {
                "Ref": "AssetParameters108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976S3Bucket727A2FAC"
              },
              "/",
              {
                "Fn::Select": [
                  0,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976S3VersionKey75C9A5D5"
                      }
                    ]
                  }
                ]
              },
              {
                "Fn::Select": [
                  1,
                  {
                    "Fn::Split": [
                      "||",
                      {
                        "Ref": "AssetParameters108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976S3VersionKey75C9A5D5"
                      }
                    ]
                  }
                ]
              }
            ]
          ]
        }
      },
      "UpdateReplacePolicy": "Delete",
      "DeletionPolicy": "Delete"
    }
  },
  "Parameters": {
    "AssetParameters108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976S3Bucket727A2FAC": {
      "Type": "String",
      "Description": "S3 bucket for asset \"108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976\""
    },
    "AssetParameters108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976S3VersionKey75C9A5D5": {
      "Type": "String",
      "Description": "S3 key for asset version \"108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976\""
    },
    "AssetParameters108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976ArtifactHashEC694637": {
      "Type": "String",
      "Description": "Artifact hash for asset \"108367430b5665d5b2b0f0c6c4393a84d9eb6d25de8b9a5b9db93e7f55683976\""
    }
  }
}
